### PR TITLE
SNIP 12 Utilities

### LINKED ISSUE #409: EIP-712 / SNIP-12
**🧐 Motivation**

[EIP-712](https://eips.ethereum.org/EIPS/eip-712) is implemented in a [draft](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/draft-EIP712.sol) in the solidity contracts. It seems like a good idea to implement in cairo as well considering that Argent-X [seems to be using it it](https://github.com/argentlabs/argent-x/discussions/14). 

**📝 Details**

I'm not entirely sure about Cairo's capabilities for this. The standard library does [support keccak](https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/cairo/common/keccak.cairo). Everything would be a felt as well.
Any guidance on how one would approach this is highly appreciated.

### PR DESCRIPTION
**Pull request by ericnordelo**
Fixes #409 

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Added entry to CHANGELOG.md 
- [ ] Tried the feature on a public network

### CODE REVIEW

**Review by andrew-fleming — COMMENTED** [review]
Excellent start, Eric! I know we're still waiting on things to become finalized, but I left some initial feedback

File: docs/modules/ROOT/pages/utilities.adoc [hunk 1]
```diff
@@ -105,6 +113,110 @@ See xref:#selectors[`openzeppelin::utils::selectors`].
...
+
+See xref:#snip12[`openzeppelin::utils::cryptography::snip12`].
+
+[.contract]
+[[NoncesComponent]]
+=== `++NoncesComponent++`
+
+```javascript
+use openzeppelin::utils::cryptography::nonces::NoncesComponent;
+```
```
[the remark this answers is not in the document]
Comment (ericnordelo) [reply docs/modules/ROOT/pages/utilities.adoc:157]: Agree, I created a cryptography section. Let me know what you think.

File: src/tests/cryptography/test_snip12.cairo [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+fn test_OffchainMessageHashImpl() {
+    let message = Message { recipient: RECIPIENT(), amount: 100, nonce: 1, expiry: 1000 };
+    let domain = StarknetDomain {
+        name: 'SNIP12Mock', version: 'v1', chain_id: 'TEST', revision: 1
+    };
+
+    starknet::testing::set_chain_id('TEST');
+
+    let expected = poseidon_hash_span(
+        array!['StarkNet Message', domain.hash_struct(), OWNER().into(), message.hash_struct()]
```
[the remark this answers is not in the document]
Comment (ericnordelo) [reply src/tests/cryptography/test_snip12.cairo:41]: I wouldn't wait for this to merge, the standard has a lot of ongoing discussions already :(

File: docs/modules/ROOT/pages/guides/snip12.adoc [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+----
+struct Message {
+    recipient: ContractAddress,
+    amount: u256,
+    nonce: felt252,
+    expiry: u64
+}
+----
+
+=== 2. Get the Message Type Hash.
```
[the remark this answers is not in the document]
Comment (ericnordelo) [reply docs/modules/ROOT/pages/guides/snip12.adoc:96]: I don't think so

File: src/utils/cryptography/interface.cairo [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts for Cairo v0.10.0 (utils/cryptography/interface.cairo)
+
+use starknet::ContractAddress;
+
+#[starknet::interface]
+trait INonces<TState> {
+    fn nonces(self: @TState, owner: ContractAddress) -> felt252;
```
Comment (martriay) [inline src/utils/cryptography/interface.cairo:8]: i think felt252 might be just fine. but I wonder if you had any thoughts on using u256

File: src/utils/cryptography/snip12.cairo [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+use starknet::{ContractAddress, get_tx_info};
+
+// selector!(
+//   "\"StarknetDomain\"(
+//    \"name\":\"shortstring\",
+//    \"version\":\"shortstring\",
+//    \"chainId\":\"shortstring\",
+//    \"revision\":\"shortstring\"
+//   )"
+// );
```
Comment (martriay) [inline src/utils/cryptography/snip12.cairo:16]: I wonder if we should just use the macro directly

File: src/utils/cryptography/snip12.cairo [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+> of OffchainMessageHash<T> {
+    fn get_message_hash(self: @T, caller: ContractAddress) -> felt252 {
+        let domain = StarknetDomain {
+            name: metadata::name(),
+            version: metadata::version(),
+            chain_id: get_tx_info().unbox().chain_id,
+            revision: 1
+        };
+        let mut state = PoseidonTrait::new();
+        state = state.update_with('StarkNet Message');
```
Comment (martriay) [inline src/utils/cryptography/snip12.cairo:62]: i know this is the standard, but the N should have been lowercased: `"Starknet Message"`

File: src/utils/cryptography/snip12.cairo (hunk 5 above)
Comment (ericnordelo) [reply src/utils/cryptography/snip12.cairo:16]: More expensive step(gas)-wise for a constant thing.

File: src/utils/cryptography/snip12.cairo (hunk 6 above)
Comment (ericnordelo) [reply src/utils/cryptography/snip12.cairo:62]: +1

File: src/utils/cryptography/interface.cairo (hunk 4 above)
Comment (ericnordelo) [reply src/utils/cryptography/interface.cairo:8]: felt252 should be cheaper than u256 even in the future if we have a smaller felt. As Ariel confirmed, I don't think fel252 will be removed, so for this case, it suits us better IMO.

File: docs/modules/ROOT/pages/utilities.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply docs/modules/ROOT/pages/utilities.adoc:217]: Looks much better!

**Review by andrew-fleming — APPROVED** [review]
I left a few tiny suggestions; otherwise, I think this is good to go (and fix changelog conflicts)! Nice work :)

**Pull request merged**